### PR TITLE
core: generalize TraceElements gatherer and add animated elements

### PR DESCRIPTION
**Pull request by adamraine**
This PR generalizes the current `TraceElements` gatherer for any metric/audit not just `largest-contentful-paint` and `cumulative-layout-shift`.  We need these changes to be able to resolve animated node ids in the trace to their label, snippet, etc.

The animated elements will be used in https://github.com/GoogleChrome/lighthouse/pull/11105 which will report non composited animations and their target nodes.  https://github.com/GoogleChrome/lighthouse/pull/11105 currently has similar changes to `TraceElements` but these will be removed.

@Beytoven's https://github.com/GoogleChrome/lighthouse/pull/11094 also resolves node ids in a new gatherer and should be able to use this updated TraceElements gatherer instead.

### CODE REVIEW

File: lighthouse-cli/test/fixtures/perf/preload_tester.js [hunk 1]
```diff
@@ -8,6 +8,7 @@
 
 document.write('<script src="/perf/level-2.js?delay=500"></script>');
 document.write('<script src="/perf/level-2.js?warning&delay=500"></script>');
+document.write('<div id="animate-me"></div>')
```
Comment (patrickhulce) [inline lighthouse-cli/test/fixtures/perf/preload_tester.js:11]: feels a bit odd to add it to the `preload.html` tester, I think `trace-elements.html` might be a more natural fit?

could just add an inline `<style>` over there with an animation

File: lighthouse-core/gather/gatherers/trace-elements.js [hunk 2]
```diff
@@ -137,6 +137,26 @@ class TraceElements extends Gatherer {
     return topFive;
   }
 
+  /**
+   * Find the node ids of elements which are animated using the Animation trace events.
+   * @param {Array<LH.TraceEvent>} mainThreadEvents
+   * @return {Array<TraceElementData>}
+   */
+  static getAnimatedElements(mainThreadEvents) {
+    const animatedElementIds = [...new Set(mainThreadEvents
```
Comment (patrickhulce) [inline lighthouse-core/gather/gatherers/trace-elements.js:146]: was the array -> set -> array an artifact of the code before?

might make more sense to avoid the unnecessary array copy and just do `for (const nodeId of idSet)` instead

File: lighthouse-core/gather/gatherers/trace-elements.js [hunk 3]
```diff
@@ -155,35 +175,42 @@ class TraceElements extends Gatherer {
...
+    const backendNodeDataMap = new Map([
+      ['largest-contentful-paint', lcpNodeId ? [{nodeId: lcpNodeId}] : []],
+      ['cumulative-layout-shift', [...clsNodeData]],
+      ['CLS/non-composited-animations', [...animatedElementData]],
+    ]);
 
-      if (response && response.result && response.result.value) {
-        traceElements.push({...response.result.value, score: backendNodeData[i].score});
+    const traceElements = [];
+    for (const [metricName, backendNodeData] of backendNodeDataMap) {
```
Comment (patrickhulce) [inline lighthouse-core/gather/gatherers/trace-elements.js:187]: @paulirish were you onboard for renaming `metricName` to something else?

`traceEventType: 'largest-contentful-paint'|'layout-shift'|'animation'` perhaps?

File: lighthouse-core/gather/gatherers/trace-elements.js [hunk 4]
```diff
@@ -155,35 +175,42 @@ class TraceElements extends Gatherer {
...
-          ${pageFunctions.getBoundingClientRectString};
-          return setAttributeMarker.call(this, '${metricName}');
-        }`,
-        returnByValue: true,
-        awaitPromise: true,
-      });
+    const backendNodeDataMap = new Map([
+      ['largest-contentful-paint', lcpNodeId ? [{nodeId: lcpNodeId}] : []],
+      ['cumulative-layout-shift', [...clsNodeData]],
+      ['CLS/non-composited-animations', [...animatedElementData]],
```
Comment (patrickhulce) [inline lighthouse-core/gather/gatherers/trace-elements.js:183]:
```suggestion
      ['cumulative-layout-shift', clsNodeData],
      ['CLS/non-composited-animations', animatedElementData],
```

these didn't seem necessary, was there a reason we need a copy?

File: lighthouse-core/test/gather/gatherers/trace-elements-test.js [hunk 5]
```diff
@@ -268,4 +292,18 @@ describe('Trace Elements gatherer - GetTopLayoutShiftElements', () => {
     const total = sumScores(result);
     expectEqualFloat(total, 2.5);
   });
+
+  it('gets animated node ids without duplicates', () => {
```
Comment (patrickhulce) [inline lighthouse-core/test/gather/gatherers/trace-elements-test.js:296]: it'd be nice to add a test that exercises the bulk of the logic we're mutating in the gatherer here with the combination of multiple nodeId types

looks like `afterPass` tests were missing in the first go around, bummer you'll have to start from scratch :/

trace-processor-test has a few good examples of using `createTestTrace` while pushing some custom events

https://github.com/GoogleChrome/lighthouse/blob/439596e9cc5e767864b2ab29ba21782246a3df1d/lighthouse-core/test/lib/tracehouse/trace-processor-test.js#L355-L363

Comment (adamraine) [reply lighthouse-core/test/gather/gatherers/trace-elements-test.js:296]: Creating an `afterPass` test will require some kind of phony `driver` to simulate `driver.resolveNodeIdToObjectId()` and `driver.sendCommand()`.

Would the solution here be to simulate these functions with hardcoded results for each node id / object id?

File: lighthouse-core/gather/gatherers/trace-elements.js (hunk 3 above)
Comment (Beytoven) [reply lighthouse-core/gather/gatherers/trace-elements.js:185]: +1 to renaming. I don't believe `metricName` makes sense any longer.

File: lighthouse-cli/test/smokehouse/test-definitions/perf/expectations.js [hunk 6]
```diff
@@ -156,6 +156,76 @@ module.exports = [
     },
   },
   {
+    artifacts: {
+      TraceElements: [
+        {
+          'metricName': 'largest-contentful-paint',
+          'devtoolsNodePath': '0,HTML,1,BODY,1,DIV,0,IMG',
```
Comment (Beytoven) [inline lighthouse-cli/test/smokehouse/test-definitions/perf/expectations.js:163]: We should leave out `devtoolsNodePath` in the smoke test to avoid the test becoming flaky. Also, I believe `nodeId` are run specific and I don't think we can guarantee their value every time.

File: lighthouse-core/gather/gatherers/trace-elements.js [hunk 7]
```diff
@@ -150,40 +170,45 @@ class TraceElements extends Gatherer {
...
-          ${pageFunctions.getBoundingClientRectString};
-          return setAttributeMarker.call(this, '${metricName}');
-        }`,
-        returnByValue: true,
-        awaitPromise: true,
-      });
+    const backendNodeDataMap = new Map([
+      ['largest-contentful-paint', lcpNodeId ? [{nodeId: lcpNodeId}] : []],
+      ['cumulative-layout-shift', clsNodeData],
+      ['CLS/non-composited-animations', animatedElementData],
```
Comment (Beytoven) [inline lighthouse-core/gather/gatherers/trace-elements.js:181]: Can we just drop the CLS prefix and name it `non-composited-animations`?

File: lighthouse-core/test/gather/gatherers/trace-elements-test.js (hunk 5 above)
Comment (patrickhulce) [reply lighthouse-core/test/gather/gatherers/trace-elements-test.js:296]: Correct, you'll need to stub out those. We haven't done a very good job in the past of doing this in gatherer tests since most of them were written before we had a framework for mocking the driver well, but since then we have a pretty nice mockCommand approach in the driver tests you can follow

https://github.com/GoogleChrome/lighthouse/blob/92b7c7642ad90f9cfd9231973dee7ef04675138e/lighthouse-core/test/gather/driver-test.js#L137

you would hardcode the node id and object id for these. I'm not looking for anything enormous or complex, just *something* that checks it tries to find each of the 3 node types and merges them with the correct labels, so something roughly like

```js
  it('identifies the node ID', async () => {
    connectionStub.sendCommand = createMockSendCommandFn()
      .mockResponse('DOM.resolveNode', {object: {objectId: 123}})
      .mockResponse('Runtime.callFunctionOn', {result: {value: {all: 'the', properties: 1}}})
      .mockResponse('DOM.resolveNode', {object: {objectId: 124}})
      .mockResponse('Runtime.callFunctionOn', {result: {value: {all: 'the', properties: 1}}})
      .mockResponse('DOM.resolveNode', {object: {objectId: 125}})
      .mockResponse('Runtime.callFunctionOn', {result: {value: {all: 'the', properties: 1}}});

    const fakeTrace = createFakeTestTrace();
    const driver = new Driver(connectionStub);
    const result = await TraceElementsGatherer.afterPass({driver}, {trace: fakeTrace});
  });
```

File: lighthouse-core/gather/gatherers/trace-elements.js (hunk 7 above)
Comment (patrickhulce) [reply lighthouse-core/gather/gatherers/trace-elements.js:181]: How do you feel about my proposal in https://github.com/GoogleChrome/lighthouse/pull/11138#discussion_r458332330 @Beytoven ?

File: lighthouse-core/gather/gatherers/trace-elements.js [hunk 8]
```diff
@@ -137,6 +137,26 @@ class TraceElements extends Gatherer {
...
+   */
+  static getAnimatedElements(mainThreadEvents) {
+    const animatedElementIds = new Set(mainThreadEvents
+      .filter(e => e.name === 'Animation' && e.ph === 'b')
+      .map(e => this.getNodeIDFromTraceEvent(e)));
+
+    /** @type Array<TraceElementData> */
+    const animatedElementData = [];
+    for (const nodeId of animatedElementIds) {
+      if (nodeId !== undefined) {
```
Comment (connorjclark) [inline lighthouse-core/gather/gatherers/trace-elements.js:153]: FWIW `node_id`, if present and a number, will always be > 0. https://source.chromium.org/chromium/chromium/src/+/master:components/ui_devtools/ui_element.cc;l=133;drc=0ee4438ca83448bca93cb11bc3012856d29303dd;bpv=0;bpt=1

so could either:

* tack on a filter(Boolean) to the end of setting `animatedElementIds`.
* or use `nodeId && animatedElementData.push({nodeId})` in the loop

File: lighthouse-cli/test/smokehouse/test-definitions/perf/expectations.js [hunk 9]
```diff
@@ -156,6 +156,67 @@ module.exports = [
...
+          'score': 0.05873307291666667,
+        },
+        {
+          'traceEventType': 'cumulative-layout-shift',
+          'selector': 'body > div#late-content > span',
+          'nodeLabel': 'Sorry!',
+          'snippet': '<span>',
+          'boundingRect': {
+            'top': 426,
+            'bottom': 444,
```
Comment (adamraine) [inline lighthouse-cli/test/smokehouse/test-definitions/perf/expectations.js:197]: Is `boundingRect.bottom` flaky?  The smoke tests keep failing on this.

Comment (patrickhulce) [reply lighthouse-cli/test/smokehouse/test-definitions/perf/expectations.js:197]: could replace this with
```suggestion
            'bottom': '444 +/- 1',
```
 
and same for height?

File: lighthouse-core/test/gather/gatherers/trace-elements-test.js [hunk 10]
```diff
@@ -268,4 +318,107 @@ describe('Trace Elements gatherer - GetTopLayoutShiftElements', () => {
...
+        },
+      ])
+    );
+    trace.traceEvents.push(makeAnimationTraceEvent('1', 5));
+    trace.traceEvents.push(makeLCPTraceEvent(6));
+
+    const gatherer = new TraceElementsGatherer();
+    const result = await gatherer.afterPass({driver}, {trace});
+
+    expect(result).toEqual([
```
Comment (patrickhulce) [inline lighthouse-core/test/gather/gatherers/trace-elements-test.js:408]: beautiful 😍 nicely done!!

File: lighthouse-core/test/results/sample_v2.json [hunk 11]
```diff
@@ -1719,90 +1719,24 @@
       "title": "Largest Contentful Paint element",
       "description": "This is the largest contentful element painted within the viewport. [Learn More](https://web.dev/lighthouse-largest-contentful-paint/)",
       "score": null,
-      "scoreDisplayMode": "informative",
-      "displayValue": "1 element found",
+      "scoreDisplayMode": "notApplicable",
+      "displayValue": "0 elements found",
```
Comment (patrickhulce) [inline lighthouse-core/test/results/sample_v2.json:1723]: I think you'll need to manually update the artifacts with the new traceEventType approach

File: types/artifacts.d.ts [hunk 12]
```diff
@@ -489,14 +489,17 @@ declare global {
         boundingRect: Rect;
       }
 
+      export type TraceEventType = 'largest-contentful-paint'|'cumulative-layout-shift'|'animation';
```
Comment (patrickhulce) [inline types/artifacts.d.ts:492]: let's just do this inline so we don't have a `LH.Artifacts.TraceEventType` floating around when it's specific to our trace elements gatherer :)

**Comment by patrickhulce** [comment]
@brendankenny do you have an interest in reviewing this? I can takeover the auto-assignment from github if you've got enough on your plate :)

File: lighthouse-cli/test/smokehouse/test-definitions/perf/expectations.js (hunk 9 above)
Comment (adamraine) [reply lighthouse-cli/test/smokehouse/test-definitions/perf/expectations.js:197]: Should this apply to every trace element or just the ones which are confirmed to be flaky?

Comment (patrickhulce) [reply lighthouse-cli/test/smokehouse/test-definitions/perf/expectations.js:197]: I would only do it to ones that seem flaky. I'm guessing this is failing because of OS-specific font line-height rendering

Comment (brendankenny) [reply lighthouse-cli/test/smokehouse/test-definitions/perf/expectations.js:197]: > Should this apply to every trace element or just the ones which are confirmed to be flaky?

ideally it should be as few as are needed (or OTOH if it's a value that's known to vary by design, then it makes sense to use bounds for all of them).

`boundingRect` should be deterministic if the test page is deterministic. Is there a reason this element is flaky and we could fix in the test fixture instead? If not, maybe leave a comment about why it might vary so the next person editing this file doesn't just think we were lazy in accepting a flaky test? :)

Comment (brendankenny) [reply lighthouse-cli/test/smokehouse/test-definitions/perf/expectations.js:197]: I was too slow :)

> I'm guessing this is failing because of OS-specific font line-height rendering, could alternatively set a fixed height on the element instead

since there's nothing testing OS-specific font rendering, this fix makes sense to me to tighten up the testing we do care about

**Comment by brendankenny** [comment]
> @brendankenny do you have an interest in reviewing this? I can takeover the auto-assignment from github if you've got enough on your plate :)

everything looks like it's well taken care of so I'll leave it to you :)

File: lighthouse-cli/test/fixtures/perf/delayed-element.js [hunk 13]
```diff
@@ -33,6 +33,7 @@ setTimeout(() => {
   const top = document.getElementById('late-content');
   top.appendChild(imgEl);
   top.appendChild(textEl);
+  top.setAttribute('height', 18); // height can be flaky so we set it manually
```
Comment (adamraine) [inline lighthouse-cli/test/fixtures/perf/delayed-element.js:36]: Setting the height changed the layout shift results so I set the height here.  Unfortunately this has not fixed the flaky smoke test.

@patrickhulce did you have another way of setting the height in mind?

Comment (patrickhulce) [reply lighthouse-cli/test/fixtures/perf/delayed-element.js:36]: `height` is only a special attribute for specific elements, you should use `.style.height = '18px'` for a `<div>`

but that's also not the element we want to set a fixed height for, isn't it? I thought the `span` *inside* `top` is the one that's the one with a flaky height?

Comment (adamraine) [reply lighthouse-cli/test/fixtures/perf/delayed-element.js:36]: Ah yes your right.  I must have misread the error.

Comment (adamraine) [reply lighthouse-cli/test/fixtures/perf/delayed-element.js]:
```
{
          "traceEventType": "layout-shift",
          "devtoolsNodePath": "0,HTML,1,BODY,1,DIV,1,SPAN",
          "selector": "body > div#late-content > span",
          "nodeLabel": "Sorry!",
          "snippet": "<span style=\"height: 18px;\">",
          "boundingRect": {
            "top": 426,
            "bottom": 443,
            "left": 8,
            "right": 49,
            "width": 41,
            "height": 17
          },
          "score": 0.003006266276041667,
          "nodeId": 13
}
```
The height is clearly set to 18px in the snippet but the `boundingRect.height` shows 17. Am I missing something?

Comment (patrickhulce) [reply lighthouse-cli/test/fixtures/perf/delayed-element.js]: `height` doesn't apply to `display: inline` elements. It's a `<span>` so the default display is `inline`.

You'll need to switch it to a `<div>` or apply `display: inline-block` or `display: block`

Comment (adamraine) [reply lighthouse-cli/test/fixtures/perf/delayed-element.js]: Looks like smoke tests are passing now.  Big thanks!

File: lighthouse-core/gather/gatherers/trace-elements.js [hunk 14]
```diff
@@ -150,40 +168,46 @@ class TraceElements extends Gatherer {
...
-          ${pageFunctions.getBoundingClientRectString};
-          return setAttributeMarker.call(this, '${metricName}');
-        }`,
-        returnByValue: true,
-        awaitPromise: true,
-      });
+    /** @type Map<LH.Artifacts.TraceEventType, {nodeId: number, score?: number}[]> */
+    const backendNodeDataMap = new Map([
+      ['largest-contentful-paint', lcpNodeId ? [{nodeId: lcpNodeId}] : []],
+      ['cumulative-layout-shift', clsNodeData],
```
[the remark this answers is not in the document]
Comment (Beytoven) [reply lighthouse-core/gather/gatherers/trace-elements.js]: I'm fine with renaming here as well. It was named `cumulative-layout-shift` to correspond to `metricTag` and it made sense to name it after the metric it was tied to. However, we're using `traceEventType` instead and its more appropriate to call it just `layout-shift` as that is the type of the trace event anyways.

TL;DR: I agree @patrickhulce

**Comment by adamraine** [comment]
I am working on changes which will resolve the animation ID to its animation name using DevTools functions `Animation.resolveAnimation` and `Runtime.getProperties`. The plan is to resolve the animation IDs in this gatherer and add a list of `{animationId, animationName}` to each animated node in the `TraceElements` artifact.

Is this a good approach or should the animation names be resolved in a separate gatherer?

**Comment by adamraine** [comment]
@patrickhulce Before we give this the green light, what do you think of adding code to resolve the animation name in this PR?

File: lighthouse-core/gather/gatherers/trace-elements.js [hunk 15]
```diff
@@ -137,6 +134,24 @@ class TraceElements extends Gatherer {
...
+   */
+  static getAnimatedElements(mainThreadEvents) {
+    const animatedElementIds = new Set(mainThreadEvents
+      .filter(e => e.name === 'Animation' && e.ph === 'b')
+      .map(e => this.getNodeIDFromTraceEvent(e)));
+
+    /** @type Array<TraceElementData> */
+    const animatedElementData = [];
+    for (const nodeId of animatedElementIds) {
+      nodeId && animatedElementData.push({nodeId});
```
[the remark this answers is not in the document]
Comment (adamraine) [reply lighthouse-core/gather/gatherers/trace-elements.js:150]: Good question.  I don't remember what compelled me to use this syntax instead of `if(nodeId) { ... }`.  I was probably in the habit after using short circuits for trace event values (e.g. `event.args && event.args.data && event.args.data.compositeFailed`).

**Comment by patrickhulce** [comment]
> what do you think of adding code to resolve the animation name in this PR?

Bold move to propose a scope change to your own PR at the 11th hour ;) Hahaha, I'm not sure how much is involved there. Is it just plucking the property from out of the trace event? Or is there more deduping with extra logic and tests?

If there's extra logic, I'd suggest merging this and handling that in #11105 or a separate PR. If it's a two-line, "expose this property too" type of change, then go for it :)

**Comment by adamraine** [comment]
I already started on some code and it's definitely not a two-liner.

Gonna go ahead and merge this one and start another PR.
